analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotTestThisWithIsOperator.cs

```diff
@@ -19,38 +19,111 @@
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, expression.GetLocation()));
-                    }
-                },
-                SyntaxKind.IsExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(AnalyzeIsExpression, SyntaxKind.IsExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(AnalyzeIsPatternExpression, SyntaxKindEx.IsPatternExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(AnalyzeSwitchExpression, SyntaxKindEx.SwitchExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(AnalyzeSwitchStatement, SyntaxKind.SwitchStatement);
         }
+
+        private static void AnalyzeIsExpression(SyntaxNodeAnalysisContext context)
+        {
+            if (((BinaryExpressionSyntax)context.Node).Left.RemoveParentheses() is ThisExpressionSyntax)
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.Node.GetLocation()));
+            }
+        }
+
+        private static void AnalyzeIsPatternExpression(SyntaxNodeAnalysisContext context)
+        {
+            var isPatternExpression = (IsPatternExpressionSyntaxWrapper)context.Node;
+            if (isPatternExpression.Expression.RemoveParentheses() is ThisExpressionSyntax
+                && ContainsTypeCheckInPattern(isPatternExpression))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.Node.GetLocation()));
+            }
+        }
+
+        private static void AnalyzeSwitchStatement(SyntaxNodeAnalysisContext context)
+        {
+            var switchStatement = (SwitchStatementSyntax)context.Node;
+            if (switchStatement.Expression.RemoveParentheses() is ThisExpressionSyntax
+                && (ContainsTypeCheckInPattern(switchStatement)
+                    || ContainsTypeCheckInCaswSwitchLabel(switchStatement)))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.Node.GetLocation()));
+            }
+        }
+
+        private static void AnalyzeSwitchExpression(SyntaxNodeAnalysisContext context)
+        {
+            var switchExpression = (SwitchExpressionSyntaxWrapper)context.Node;
+            if (switchExpression.GoverningExpression.RemoveParentheses() is ThisExpressionSyntax
+                && ContainsTypeCheckInPattern(switchExpression))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.Node.GetLocation()));
+            }
+        }
+
+        private static bool ContainsTypeCheckInCaswSwitchLabel(SwitchStatementSyntax switchStatement) =>
+            switchStatement.Sections.Any(x => x.Labels.Any(x => x is CaseSwitchLabelSyntax caseSwitchLabel
+                                                                && caseSwitchLabel.Value.IsKind(SyntaxKind.IdentifierName)));
+
+        private static bool ContainsTypeCheckInPattern(SyntaxNode syntaxNode) =>
```

Comment:
This is called on the entire switch node. If there would be nested type pattern check inside the returned value, this looks like it would raise a FP. Let's add test case.

I think the `DescendantNodes` should have stopping condition to avoid nesting inside the returned value/block/section.

Reply:
Did you take any action here about the stopping condition, @csaba-sagi-sonarsource ? I see you didn't reply to this comment and didn't change the code, either?

Reply:
The stopping conditions are the same but I added some logic and test cases and now we do not raise FPs.